cmd/tpm-manager/tpm-manager.go

```diff
@@ -0,0 +1,507 @@
+	}
+	node.ObjectMeta.Annotations[LogState] = log
+}
+
+func verifyNode(node *api.Node) error {
+	address, err := nodeutil.GetNodeHostIP(node)
+	if err != nil {
+		return err
+	}
+	host := fmt.Sprintf("%s:23179", address.String())
+	tpmdata, err := manager.tpmhandler.Get(host, manager.allowUnknown)
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: Unable to obtain TPM data for node %s: %v", address.String(), err)
+	}
+	quote, log, err := tpm.Quote(tpmdata)
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: Unable to obtain TPM quote for node %s: %v", address.String(), err)
+	}
+
+	err = tpm.ValidateLog(log, quote)
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: TPM event log does not match quote for node %s", address.String())
+	}
+
+	// Don't handle this error immediately - we want to update the annotations even if validation failed
+	logstate, err := tpm.ValidatePCRs(log, quote, manager.pcrConfigs)
+	jsonlog, jsonerr := json.Marshal(logstate)
+
+	if jsonerr == nil {
+		updateAnnotations(node, string(jsonlog))
+	}
+
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: Unable to validate quote for node %s", address.String())
+	}
+
+	// If we've got this far then the node is trustworthy
+	trustNode(node)
+	return nil
+}
+
+func verifyAndUpdate(node *api.Node) {
+	verifyNode(node)
+
+	// The state that the node will be updated to
+	newstate := isTrusted(node)
+	newnode, err := manager.client.Nodes().Get(node.Name)
+	if err != nil {
+		glog.Errorf("Unable to obtain node state for %s: %v", node.Name, err)
+		return
+	}
+	if newnode.ObjectMeta.Annotations == nil {
+		newnode.ObjectMeta.Annotations = make(map[string]string)
+	}
+	currenttime := time.Now().Unix()
+
+	// If we're transitioning state, update the metadata
+	if isTrusted(newnode) != newstate {
+		if newstate == false {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = strconv.FormatInt(currenttime, 10)
+			newnode.ObjectMeta.Annotations[TrustedSince] = ""
+		} else {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = ""
+			newnode.ObjectMeta.Annotations[TrustedSince] = strconv.FormatInt(currenttime, 10)
+		}
+	}
+
+	newnode.ObjectMeta.Annotations[ValidationTime] = strconv.FormatInt(currenttime, 10)
+
+	// Ensure that the new node is tainted appropriately
+	if newstate == true {
+		trustNode(newnode)
+	} else {
+		invalidateNode(newnode)
+	}
+	newnode.ObjectMeta.Annotations[LogState] = node.ObjectMeta.Annotations[LogState]
+	newnode, err = manager.client.Nodes().Update(newnode)
+	if err != nil {
+		glog.Errorf("Unable to update node state for %s: %v", node.Name, err)
+		return
+	}
+}
+
+func verifyAllNodes() {
+	nodes, err := manager.client.Nodes().List(api.ListOptions{})
+	if err != nil {
+		glog.Errorf("Unable to obtain list of nodes")
+		return
+	}
+	for _, node := range nodes.Items {
+		verifyAndUpdate(&node)
```

Comment:
Seems somewhat inefficient to first retrieve all nodes, then again retrieve each node individually from api again in the verifyAndUpdate function.